pkg/certificate/cert_manager/provider.go

```diff
@@ -0,0 +1,254 @@
+package cert_manager
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"strings"
+	"time"
+
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	interfaces "go.etcd.io/etcd-operator/pkg/certificate/interfaces"
+)
+
+type Provider struct {
+	client.Client
+}
+
+func New(c client.Client) *Provider {
+	return &Provider{
+		c,
+	}
+}
+
+// createCertificate creates a cert-manager Certificate resource in the specified namespace.
+// DNSNames and IPAddresses if not user-defined, will be set to default value in runtime:
+// fmt.Sprintf("%s-%d.%s.%s.svc.cluster.local", ec.Name, index, ec.Name, ec.Namespace)
+// returns an error if the Certificate resource cannot be created.
+func (cm *Provider) createCertificate(
+	ctx context.Context,
+	secretName string,
+	namespace string,
+	cfg *interfaces.Config) error {
+	issuerName := cfg.ExtraConfig["issuerName"].(string)
+	issuerKind := cfg.ExtraConfig["issuerKind"].(string)
+
+	certificateResource := &certmanagerv1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: namespace,
+		},
+		Spec: certmanagerv1.CertificateSpec{
+			SecretName:  secretName,
+			DNSNames:    cfg.AltNames.DNSNames,
+			IPAddresses: strings.Fields(strings.Trim(fmt.Sprint(cfg.AltNames.IPs), "[]")),
+			IssuerRef: cmmeta.ObjectReference{
+				Name: issuerName,
+				Kind: issuerKind,
+			},
+			Duration: &metav1.Duration{Duration: cfg.ValidityDuration},
```

Comment:
why I do not see some fields (i.e. CommonName, Organization) in the Config?

Reply:
If there is something wrong on the cert-manager side, then it may get stuck on creating the secret (the reason will be "Pending" for a long long time), we need to think about how to resolve it.
- Solution 1: Let users to manually resolve it ( because the cert-manager is deployed by users), and need human intervention.When users cancel/remove the certificate resource, etcd-operator will automatically recreate it.
- Solution 2: etcd-operator automaticaly cancel it if the secret isn't ready for a long time (a timeout parameter is needed from users, and have a default value), and recreate the certificate resource.

Reply:
> why I do not see some fields (i.e. CommonName, Organization) in the Config?

Updated the missing fields.